File: src/TheWorld/wwwroot/js/tripsController.js
//TripsController.js

(function () {
    "use strict";

    //Getting the Module (existing)
    angular.module("app-trips")
    .controller("tripsController", tripsController);


    function tripsController($http) {
        var vm = this;
        vm.trips = [];

        vm.newTrip = {};

        vm.errorMessage = "";
        vm.isBusy = true;

        $http.get("/api/trips")
            .then(function(response) {
                    //success 
                    angular.copy(response.data, vm.trips);
                },
                function(error) {
                    //failure
                    vm.errorMessage = "Failed to load data :" + error;
                })
        .finally(function() {
            vm.isBusy = false;
            });

        vm.addTrip = function() {
            vm.isBusy = true;
            vm.errorMessage = "";
            $http.post("/api/trips", vm.newTrip)
                .then(function(response) {
                    vm.trips.push(response.data);
                    vm.newTrip = {};
                    },
                    function(error) {
                        vm.errorMessage = "Failed to save new trip :" + error;
                    })
                .finally(function() {
                    vm.isBusy = false;
                });

        }
    }
})();

File: src/TheWorld/Controllers/Api/TripsController.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using TheWorld.Models;
using TheWorld.ViewModels;

namespace TheWorld.Controllers.Api
{
    [Route("api/trips")]
    [Authorize]
    public class TripsController : Controller
    {
        private readonly IWorldRepository _repository;
        private readonly ILogger<TripsController> _logger;

        public TripsController(IWorldRepository repository,ILogger<TripsController> logger )
        {
            _repository = repository;
            _logger = logger;
        }

        [HttpGet("")]

        public IActionResult Get()
        {
            try
            {
                var result = _repository.GetTripsByUserName(this.User.Identity.Name);
                return Ok(Mapper.Map<IEnumerable<TripViewModel>>(result));
            }
            catch (Exception ex)
            {
                _logger.LogError($"Failed to get all trips {ex}");
                return BadRequest("Error occured");
            }

        }

        [HttpPost("")]
        public async Task<IActionResult> Post([FromBody]TripViewModel theTrip)
        {
            if (ModelState.IsValid)
            {
                var newTrip = Mapper.Map<Trip>(theTrip);
                newTrip.Username = User.Identity.Name;
                _repository.AddTrip(newTrip);

                if (await _repository.SaveChangesAsynch())
                {
                    return Created($"api/trips/{theTrip.Name}", Mapper.Map<TripViewModel>(newTrip));
                }
            }
            return BadRequest("Failed to save changes to the database");

        }
    }
}
